### PR TITLE
Clean up some usages of LowLevelList<T>

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Digging deep of the history of `LowLevelList<T>`, I've found that most usages came with the initial population of corert repo. Now the structure of NativeAOT changed a lot and we have `List<T>` and `Dictionary<TKey, TValue>` in the CoreLib.

This PR addresses the cases I have most confident, in two categories: creating arrays, where `ArrayBuilder<T>` should have better performance and less metadata surface, and I've already seen its usage; interface implementing cases in reflection access, which should be less performance critical, and uses similar metadata surface with `List<T>`.

Is there any instruction to measure the size impact?

Before continuing on more cases, I want to check the capabilities of naot toolchain these days:
- There are `ForceDictionaryLookupsAttribute`, which hints some special work to do within `TypeLoader` itself and prevents the usage of generic `List<T>`. However I can't find any toolchain getting this attribute. Is it still relevant?
- Would unused members have size/metadata impact, especially the virtual ones? Is the impact less (code sharing) or more (generic dictionary) for reference types?

### CODE REVIEW

**Comment by MichalStrehovsky** [comment]
> There are ForceDictionaryLookupsAttribute, which hints some special work to do within TypeLoader itself and prevents the usage of generic List<T>. However I can't find any toolchain getting this attribute. Is it still relevant?

This one is relevant if we ever do https://github.com/dotnet/runtime/issues/85184#issuecomment-1545059247 - TypeLoader cannot call into generic code that doesn't have this attribute, and we'd not want to put `ForceDictionaryLookups` on `List<T>`. I don't know if we'll ever get to that.

> Would unused members have size/metadata impact, especially the virtual ones? Is the impact less (code sharing) or more (generic dictionary) for reference types?

Hard to say in general terms. I've kicked off a size validation job at https://github.com/MichalStrehovsky/rt-sz/issues/50, it should post results in 20 minutes or so. It could be an improvement, but it could also be a regression. I don't expect anything major either way.

**Comment by huoyaoyuan** [comment]
Ah reflection is broken. Some replacement was incorrect.

I'd expect the result to be more positive with more existing usage of `List<T>` in application.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/General/Helpers.cs [hunk 1]
```diff
@@ -151,22 +151,26 @@ public static BinderBundle ToBinderBundle(this Binder binder, BindingFlags invok
...
+            {
+                // This is here for desktop compatibility. ICustomAttribute.GetCustomAttributes() normally returns an array of the
+                // exact attribute type requested except in two cases: when the passed in type is an open type and when
+                // it is a value type. In these two cases, it returns an array of type Object[].
+                return attributes.ToArray();
+            }
+            else
+            {
+                object[] result = (object[])Array.CreateInstance(actualElementType, attributes.Count);
+                attributes.AsSpan(0).CopyTo(result);
```
Comment (huoyaoyuan) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/General/Helpers.cs:171]: `CopyTo(Span)` can't be used when destination is covariant. Unlike `CopyTo(Array)` which does type test on each access, span does type test on construction and rejects for covariant case.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/General/MetadataReaderExtensions.NativeFormat.cs [hunk 2]
```diff
@@ -79,20 +79,21 @@ internal static Type[] GetCustomModifiers(this Handle handle, MetadataReader rea
...
                     Type customModifier = modifiedType.ModifierType.Resolve(reader, typeContext).ToType();
-                    customModifiers.Insert(0, customModifier);
+                    customModifiers.Add(customModifier);
                 }
 
                 handle = modifiedType.Type;
                 handleType = handle.HandleType;
             }
             while (handleType == HandleType.ModifiedType);
+            customModifiers.AsSpan(0).Reverse();
```
Comment (PaulusParssinen) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/General/MetadataReaderExtensions.NativeFormat.cs:96]: When I added the `ArrayBuilder<T>.AsSpan(int start)` in https://github.com/dotnet/runtime/pull/102394, I did not limit the returned `Span<T>` to the only used portion, so I believe this this `Reverse()` is currently reversing the entire underlying `ArrayBuilder<T>._items` array 😅 

I think we need to update `ArrayBuilder<T>.AsSpan(int start)` to slice the only used portion in order for this to do the correct thing.

(This concern was already raised in https://github.com/dotnet/runtime/pull/102394/files#r1605898748)

Comment (huoyaoyuan) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/General/MetadataReaderExtensions.NativeFormat.cs:96]: Digging further, it also came from the early era of CoreRT.
Now we have the more-optimized version of ArrayBuilder in CoreLib. We can borrow some optimizations there. I'm a little concerned about size optimization to directly use that version, especially calls to the monster `Array.Copy`.

File: src/coreclr/tools/Common/System/Collections/Generic/ArrayBuilder.cs [hunk 3]
```diff
@@ -46,7 +46,7 @@ public void Add(T item)
         }
 
 #if NET
-        public readonly Span<T> AsSpan(int start) => _items.AsSpan(start);
+        public readonly Span<T> AsSpan(int start) => _items.AsSpan(start, _count);
```
Comment (PaulusParssinen) [inline src/coreclr/tools/Common/System/Collections/Generic/ArrayBuilder.cs:49]: I believe this should be:

```suggestion
        public readonly Span<T> AsSpan(int start) => _items.AsSpan(start, _count - start);
```

Comment (huoyaoyuan) [reply src/coreclr/tools/Common/System/Collections/Generic/ArrayBuilder.cs]: Yes I've also caught this.

**Comment by huoyaoyuan** [comment]
Now it's all positive improvement. I think we should take step ahead and take further optimizations of `ArrayBuilder<T>` in follow-up.

**Review by jkotas — APPROVED** [review]

**Comment by huoyaoyuan** [comment]
The event listener tests were also failing in previous outerloop runs. I can't find any active issue or test skipping logging this.

**Review by MichalStrehovsky — APPROVED** [review]
Thank you! Nice improvement!

**Pull request merged**